plugins/optimization-detective/site-health.php

```diff
@@ -0,0 +1,304 @@
+ *
+ * @since n.e.x.t
+ * @access private
+ *
+ * @param array<string, mixed>|WP_Error $response REST API response.
+ * @return array{label: string, status: string, badge: array{label: string, color: string}, description: string, actions: string, test: string} Result.
+ */
+function od_compose_site_health_result( $response ): array {
+	$common_description_html = '<p>' . wp_kses(
+		sprintf(
+			/* translators: %s is the REST API endpoint */
+			__( 'To collect URL Metrics from visitors the REST API must be available to unauthenticated users. Specifically, visitors must be able to perform a <code>POST</code> request to the <code>%s</code> endpoint.', 'optimization-detective' ),
+			'/' . OD_REST_API_NAMESPACE . OD_URL_METRICS_ROUTE
+		),
+		array( 'code' => array() )
+	) . '</p>';
+
+	$result = array(
+		'label'       => __( 'Optimization Detective\'s REST API endpoint is available', 'optimization-detective' ),
+		'status'      => 'good',
+		'badge'       => array(
+			'label' => __( 'Optimization Detective', 'optimization-detective' ),
+			'color' => 'blue',
+		),
+		'description' => $common_description_html . '<p><strong>' . esc_html__( 'This appears to be working properly.', 'optimization-detective' ) . '</strong></p>',
+		'actions'     => '',
+		'test'        => 'optimization_detective_rest_api',
+	);
+
+	$error_label            = __( 'Optimization Detective\'s REST API endpoint is unavailable', 'optimization-detective' );
+	$error_description_html = '<p>' . esc_html__( 'You may have a plugin active or server configuration which restricts access to logged-in users. Unauthenticated access must be restored in order for Optimization Detective to work.', 'optimization-detective' ) . '</p>';
+
+	if ( is_wp_error( $response ) ) {
+		$result['status']      = 'critical';
+		$result['label']       = $error_label;
+		$result['description'] = $common_description_html . $error_description_html . '<p>' . wp_kses(
+			sprintf(
+				/* translators: %s is the error code */
+				__( 'The REST API responded with the error code <code>%s</code> and the following error message:', 'optimization-detective' ),
+				esc_html( (string) $response->get_error_code() )
+			),
+			array( 'code' => array() )
+		) . '</p><blockquote>' . esc_html( $response->get_error_message() ) . '</blockquote>';
+	} else {
+		$code    = wp_remote_retrieve_response_code( $response );
+		$message = wp_remote_retrieve_response_message( $response );
+		$body    = wp_remote_retrieve_body( $response );
+		$data    = json_decode( $body, true );
+
+		$required_args = array( 'slug', 'current_etag', 'hmac', 'url', 'viewport', 'elements' );
+		$is_expected   = (
+			400 === $code &&
+			isset( $data['code'], $data['data']['params'] ) &&
+			'rest_missing_callback_param' === $data['code'] &&
+			is_array( $data['data']['params'] ) &&
+			count( array_intersect( $data['data']['params'], $required_args ) ) === count( $required_args )
```

Comment:
I already flagged this in https://github.com/WordPress/performance/pull/1762#discussion_r1913459046, which was then resolved, but now this code has come back, which I think is overkill and a code smell because we'd have to keep in mind to manually maintain this list here, should these change elsewhere.

Based on that, checking that there are missing parameters alone should be sufficient, we don't need to check for the exact parameters missing.

Reply:
That's fair. Removed in ddc6164d72ff7ceee7964cea3310cc6f403ab8b3.